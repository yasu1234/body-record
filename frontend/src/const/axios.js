import axios from "axios";
import Cookies from "js-cookie";

const baseURL = import.meta.env.VITE_APP_API_BASE;

const axiosInstance = axios.create({
  baseURL,
  headers: {
    "Content-Type": "application/json",
  },
});

axiosInstance.interceptors.request.use((config) => {
  config.headers["access-token"] = Cookies.get("accessToken");
  config.headers["client"] = Cookies.get("client");
  config.headers["uid"] = Cookies.get("uid");

  return config;
});

export default axiosInstance;
