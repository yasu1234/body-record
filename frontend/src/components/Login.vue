<script setup>
import axios from 'axios'
import { ref } from 'vue'
import Cookies from 'js-cookie'

import Header from './Header.vue'

const email = ref('')
const password = ref('')

const login = async () => {
  try {
    const res = await axios.post(import.meta.env.VITE_APP_API_BASE + '/api/v1/auth/sign_in', {
      email: email.value,
      password: password.value
    })
    Cookies.set('accessToken', response.headers["access-token"])
    Cookies.set('client', response.headers["client"])
    Cookies.set('uid', response.headers["uid"])

    router.push({ name: 'Home'})
  } catch (error) {
    console.log({ error })
  }
}
</script>

<template>
    <Header />
    <h1 class="signUpTitle">ログイン</h1>
    <div class="singUpInput">
        <form class="form" @submit.prevent="login">
            <div class="item">
                <label class="itemLabel">メールアドレス</label>
                <input id="email" type="email" v-model="email">
            </div>
            <div class="item">
                <label class="itemLabel" for="password">パスワード</label>
                <input id="password" type="password" v-model="password">
            </div>
            <div class="signUpTitle">
                <button class="loginButton">ログイン</button>
            </div>
        </form>
    </div>
</template>

<style>
.form{
   width: 100%;
   margin:0 auto;
   padding: 50px;
   box-sizing: border-box;
}

.signUpTitle {
    padding-top: 40px;
    text-align: center;
}
 
.item{
    padding-top: 40px;
    width: 50%;
    margin: 0 auto;
}

.itemLabel{
    display: block;
    text-align: left;
}

.error-message{
    width: 50%;
    margin: 0 auto;
}
 
.error-message-text{
    color: red;
    font-size: 15px;
}

.form input[type="email"],
.form input[type="password"] {
     padding: 10px;
     width: 100%;
 }
 
.form textarea{
   padding: 10px;
   height: 100px;
   font-size: 14px;
   width: 50%;
 }
 
.loginButton{
    background: #ffa500;
    color: white;
    font-size:16px;
    font-weight:bold;
    padding: 10px 50px;
}
</style>
