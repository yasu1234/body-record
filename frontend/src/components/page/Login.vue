<script setup>
import axios from 'axios'
import { ref } from 'vue'
import { useRouter } from 'vue-router'
import Cookies from 'js-cookie'
import { useField, useForm } from 'vee-validate';
import * as yup from 'yup';

import Header from '../layout/Header.vue'
import ErrorMessage from '../atom/ErrorMessage.vue'

const errorMessage = ref('');

const router = useRouter();

const handleSubmit = async () => {
  const result = await validate()
  if (result.valid) {
    login()
  }
}

const login = async () => {
  try {
    const res = await axios.post(import.meta.env.VITE_APP_API_BASE + '/api/v1/auth/sign_in', {
      email: email.value,
      password: password.value
    })
    Cookies.set('accessToken', res.headers["access-token"])
    Cookies.set('client', res.headers["client"])
    Cookies.set('uid', res.headers["uid"])

    router.push({ name: 'Home'})
  } catch (error) {
    let errorMessages = 'ログインに失敗しました\n';
    if (error.response.status === 401) {
        if (Array.isArray(error.response.data.errors)) {
            errorMessages += error.response.data.errors.join('\n');
        }
    }
    errorMessage.value = errorMessages
  }
}

const schema = yup.object({
  password: yup
    .string()
    .required("この項目は必須です")
    .min(6, "6文字以上で入力してください")
    .max(128, "128文字以下で入力してください"),
  email: yup
    .string()
    .required("この項目は必須です")
    .email("メールアドレスの形式で入力してください"),
});

const { validate } = useForm({ validationSchema: schema })
const { value: password, errorMessage: passwordError } = useField('password');
const { value: email, errorMessage: emailError } = useField('email');
</script>

<template>
    <Header />
    <h1 class="signUpTitle">ログイン</h1>
    <ErrorMessage :errorMessage="errorMessage"/>
    <div class="singUpInput">
        <form class="form" @submit.prevent="handleSubmit">
            <div class="item">
                <label class="itemLabel">メールアドレス</label>
                <input id="email" type="email" v-model="email">
                <p class="validation-error-message">{{ emailError }}</p>
            </div>
            <div class="item">
                <label class="itemLabel" for="password">パスワード</label>
                <input id="password" type="password" v-model="password">
                <p class="validation-error-message">{{ passwordError }}</p>
            </div>
            <div class="signUpTitle">
                <button class="loginButton">ログイン</button>
            </div>
        </form>
    </div>
</template>

<style>
.form{
   width: 100%;
   margin:0 auto;
   padding: 50px;
   box-sizing: border-box;
}

.signUpTitle {
    padding-top: 40px;
    text-align: center;
}
 
.item{
    padding-top: 40px;
    width: 50%;
    margin: 0 auto;
}

.itemLabel{
    display: block;
    text-align: left;
}
 
.error-message-text{
    color: red;
    font-size: 15px;
}

.form input[type="email"],
.form input[type="password"] {
     padding: 10px;
     width: 100%;
 }
 
.form textarea{
   padding: 10px;
   height: 100px;
   font-size: 14px;
   width: 50%;
 }
 
.loginButton{
    background: #ffa500;
    color: white;
    font-size:16px;
    font-weight:bold;
    padding: 10px 50px;
}
</style>
