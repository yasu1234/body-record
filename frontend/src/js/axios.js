import axios from "axios";
import Cookies from "js-cookie";

const baseURL = import.meta.env.VITE_APP_API_BASE;

const axiosInstance = axios.create({
  baseURL,
  headers: {
    'Content-Type': 'application/json',
    'X-Requested-With': 'XMLHttpRequest',
  },
});

axiosInstance.interceptors.request.use((config) => {
  config.headers["access-token"] = Cookies.get("accessToken");
  config.headers["client"] = Cookies.get("client");
  config.headers["uid"] = Cookies.get("uid");
  // config.headers.common['X-Requested-With'] = 'XMLHttpRequest';
  // config.headers['X-Requested-With'] = 'XMLHttpRequest';

  return config;
});

const setupInterceptors = (router) => {
  axiosInstance.interceptors.response.use(
    response => response,
    error => {
      // 共通のエラー処理
      if (error.response != null && error.response.status === 404) {
        router.push({ name: 'NotFound' })
      }

      // エラーをキャッチしてreturnする
      return Promise.reject(error)
    }
  )
}

export { axiosInstance, setupInterceptors }
