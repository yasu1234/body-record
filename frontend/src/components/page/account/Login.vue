<template>
  <Header />
  <Toast position="top-center" />
  <h1 class="view-title mt-7">ログイン</h1>
  <div class="login-container">
    <form class="form" @submit.prevent="handleSubmit">
      <div class="login-item">
        <label>メールアドレス</label>
        <input id="email" type="email" v-model="email" />
        <p class="validation-error-message">{{ emailError }}</p>
      </div>
      <div class="login-item">
        <label>パスワード</label>
        <PasswordText :password="password" @update-password="updatePassword" />
        <p class="validation-error-message">{{ passwordError }}</p>
      </div>
      <div class="login-button-container">
        <button class="login-button">ログイン</button>
      </div>
    </form>
  </div>
</template>

<script setup>
import { useRouter } from "vue-router";
import Cookies from "js-cookie";
import { useField, useForm } from "vee-validate";
import * as yup from "yup";
import { useToast } from "primevue/usetoast";
import { toastService } from "../../../js/toast.js";
import Toast from "primevue/toast";
import { axiosInstance, setupInterceptors } from "../../../js/axios.js";

import Header from "../../layout/Header.vue";
import PasswordText from "../../atom/PasswordText.vue";

const router = useRouter();
const toast = useToast();
const toastNotifications = new toastService(toast);
setupInterceptors(router);

const handleSubmit = async () => {
  const result = await validate();
  if (result.valid) {
    login();
  }
};

const login = async () => {
  try {
    const res = await axiosInstance.post("/api/v1/auth/sign_in", {
      email: email.value,
      password: password.value,
    });
    Cookies.set("accessToken", res.headers["access-token"], { expires: 14 });
    Cookies.set("client", res.headers["client"], { expires: 14 });
    Cookies.set("uid", res.headers["uid"], { expires: 14 });

    if (
      Cookies.get("loginRoutePath") != null &&
      !Cookies.get("loginRoutePath").includes("login") &&
      !Cookies.get("loginRoutePath").includes("accountInteroduction") &&
      !Cookies.get("loginRoutePath").includes("signup")
    ) {
      router.push({ path: Cookies.get("loginRoutePath") });
      Cookies.remove("loginRoutePath");
    } else {
      showHome();
    }
  } catch (error) {
    let errorMessages = "";
    if (error.response != null && error.response.status === 422) {
      if (Array.isArray(error.response.data.errors)) {
        errorMessages += error.response.data.errors.join("\n");
      } else {
        errorMessages = error.response.data.errors;
      }
    }
    toastNotifications.displayError("ログインに失敗しました", errorMessages);
  }
};

const schema = yup.object({
  password: yup
    .string()
    .required("この項目は必須です")
    .min(6, "6文字以上で入力してください")
    .max(128, "128文字以下で入力してください"),
  email: yup
    .string()
    .required("この項目は必須です")
    .email("メールアドレスの形式で入力してください"),
});

const { validate } = useForm({ validationSchema: schema });
const { value: password, errorMessage: passwordError } = useField(
  "password",
  ""
);
const { value: email, errorMessage: emailError } = useField("email", "");

const updatePassword = (inputPassword, passwordType) => {
  password.value = inputPassword;
};

const showHome = () => {
  router.push({ name: "Home" });
};
</script>

<style scoped>
.login-container {
  width: 500px;
  margin: 0 auto;
  padding: 20px;
  background-color: #ffffff;
  border: 1px solid #ccc;
  border-radius: 5px;
}
.login-item {
  padding-top: 40px;
  margin: 0 auto;
}
.form input[type="email"] {
  padding: 10px;
  width: 100%;
  border: 1px solid #ccc;
}
.login-button-container {
  padding-top: 40px;
  text-align: center;
}
.login-button {
  font-size: 16px;
  font-weight: bold;
  padding: 10px 50px;
}

@media screen and (max-width: 768px) {
  .login-container {
    width: auto;
    margin-left: 20px;
    margin-right: 20px;
    padding: 20px;
    background-color: #ffffff;
    border: 1px solid #ccc;
    border-radius: 5px;
  }
}
</style>
