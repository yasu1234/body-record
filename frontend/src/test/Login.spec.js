import { shallowMount, flushPromises } from '@vue/test-utils'
import axios from 'axios'
import Login from '@/components/page/Login.vue'

jest.mock('axios')
jest.mock('vue-router', () => {
    const router = {
      push: jest.fn(),
      pathname: '/',
      route: '',
      asPath: '',
      query: {},
      events: {
        on: jest.fn(),
        off: jest.fn(),
        emit: jest.fn(),
      },
    };
    return {
      useRouter: jest.fn().mockReturnValue(router),
    };
  });

  jest.mock('primevue/usetoast', () => ({
    useToast: jest.fn(() => ({
      add: jest.fn(),
      removeAll: jest.fn()
    }))
  }))

describe('SearchPage', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(Login, {
        global: {
          stubs: ["router-link", "router-view"]
        }
    })
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  it('SnapShotテスト初期表示', () => {
    expect(wrapper.element).toMatchSnapshot();
  })

  it('バリデーションエラー(メールアドレス形式違い)', async () => {
    await wrapper.find('#email').setValue('invalid-email')
    await wrapper.vm.handleSubmit()
    expect(wrapper.findAll('.validation-error-message')[0].text()).toBe('メールアドレスの形式で入力してください')
  })

  it('バリデーションエラー(メールアドレス入力なし)', async () => {
    await wrapper.vm.handleSubmit()
    expect(wrapper.findAll('.validation-error-message')[0].text()).toBe('この項目は必須です')
  })

  it('バリデーションエラー(パスワード入力なし)', async () => {
    await wrapper.vm.handleSubmit()
    expect(wrapper.findAll('.validation-error-message')[1].text()).toBe('この項目は必須です')
  })

  it('バリデーションエラー(パスワード文字数不足)', async () => {
    wrapper.vm.password = 'passw'
    await wrapper.vm.handleSubmit()
    expect(wrapper.findAll('.validation-error-message')[1].text()).toBe('6文字以上で入力してください')
  })

  it('バリデーションエラー(パスワード文字数オーバー)', async () => {
    wrapper.vm.password = 'passwordpasswordpasswordpasswordpasswordpasswordpasswordpasswordpasswordpasswordpasswordpasswordpasswordpasswordpasswordpasswordpassword'
    await wrapper.vm.handleSubmit()
    expect(wrapper.findAll('.validation-error-message')[1].text()).toBe('128文字以下で入力してください')
  })
})
