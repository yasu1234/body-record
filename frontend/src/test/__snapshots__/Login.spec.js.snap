// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login SnapShotテスト初期表示 1`] = `
<div
  data-v-app=""
>
  
  <header-stub />
  <toast-stub
    position="top-center"
  />
  <h1
    class="view-title mt-7"
  >
    ログイン
  </h1>
  <div
    class="login-container"
  >
    <form
      class="form"
    >
      <div
        class="login-item"
      >
        <label>
          メールアドレス
        </label>
        <input
          id="email"
          type="email"
        />
        <p
          class="validation-error-message"
        />
      </div>
      <div
        class="login-item"
      >
        <label>
          パスワード
        </label>
        <password-text-stub />
        <p
          class="validation-error-message"
        />
      </div>
      <div
        class="login-button-container"
      >
        <button
          class="login-button"
        >
          ログイン
        </button>
      </div>
    </form>
  </div>
  
</div>
`;
